File: src-tauri/src/cmd/source.rs
use crate::{
    api::source::{Source, Sources},
    errors::AppError,
    state::ServiceAccess,
};

#[tauri::command]
pub fn validate_source_payload(
    name: &str,
    source_type: &str,
    path: &str,
    handle: tauri::AppHandle,
) -> Result<(), AppError> {
    handle.source_ctrl(|ctrl| ctrl.validate_payload(name, source_type, path))
}

#[tauri::command]
pub fn create_source(
    name: &str,
    source_type: &str,
    path: &str,
    handle: tauri::AppHandle,
) -> Result<Source, AppError> {
    handle
        .source_ctrl(|ctrl| ctrl.create(name, source_type, path))
        .map(Source::from)
}

#[tauri::command]
pub fn get_source(id: &str, handle: tauri::AppHandle) -> Result<Source, AppError> {
    handle
        .source_ctrl(|ctrl| ctrl.get_by_id(id))
        .map(Source::from)
}

#[tauri::command]
pub fn update_source(id: &str, name: &str, handle: tauri::AppHandle) -> Result<Source, AppError> {
    handle
        .source_ctrl(|ctrl| ctrl.update_by_id(id, name))
        .map(Source::from)
}

#[tauri::command(async)]
pub fn list_sources(handle: tauri::AppHandle) -> Result<Sources, AppError> {
    handle.source_ctrl(|ctrl| ctrl.list()).map(Sources::from)
}

#[tauri::command]
pub fn delete_source(id: &str, handle: tauri::AppHandle) -> Result<(), AppError> {
    handle.source_ctrl(|ctrl| ctrl.delete(id))?;
    handle.fs_entry_ctrl(|ctrl| ctrl.delete_by_source_id(id))
}

File: src/lib/api/source.ts
import { invoke } from '@tauri-apps/api';

export enum SourceType {
	Local = 'Local'
}

export interface Source {
	id: string;
	name: string;
	path: string;
	source_type: SourceType;
}

export interface Sources {
	sources: Source[];
}

export interface ScanResults {
	entries_deleted: number;
	entries_created: number;
	thumbnails_created: number;
}

export const create = async (name: string, type: string, path: string): Promise<Source> => {
	const source = (await invoke('create_source', { name, sourceType: type, path })) as Source;

	return source;
};

export const list = async (): Promise<Sources> => {
	const sources = (await invoke('list_sources', {})) as Sources;

	return sources;
};

export const get = async (id: string): Promise<Source> => {
	const source = (await invoke('get_source', { id })) as Source;

	return source;
};

export const update = async (id: string, name: string): Promise<Source> => {
	const source = (await invoke('update_source', { id, name })) as Source;

	return source;
};

export const remove = async (id: string): Promise<Source> => {
	const source = (await invoke('delete_source', { id })) as Source;

	return source;
};

export const scan = async (id: string): Promise<ScanResults> => {
	return (await invoke('scan_source_dir', { sourceId: id })) as ScanResults;
};
